<h1><center>Laboratorio 3: El Pandas no muerde 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos - Primavera 2024</strong></center>

### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Eduardo Moya
- Ayudantes: Nicolás Ojeda, Melanie Peña, Valentina Rojas

### Equipo: **SUPER IMPORTANTE - notebooks sin nombre no serán revisados**

- Nombre de alumno 1: Francisca Ruiz
- Nombre de alumno 2: Valentina Zúñiga


### **Link de repositorio de GitHub:** [Repositorio](https://github.com/vazuniga1/Laboratorios/tree/lab-3)

## Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 6 días de plazo con descuento de 1 punto por día. Entregas Jueves a las 23:59.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquier material del curso que estimen conveniente.

### Temas a tratar

- Introducción al manejo de datos tabulares por medio de la libreria `pandas`.

### Objetivos principales del laboratorio

- Entender, aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas` para manejar datos tabulares.
- Aprender a utilizar la documentación de pandas (y en general, de cualquier librería de python) como también a buscar soluciones en la web.

> **Nota**: El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

### Librerias utiles para el Lab 😸

In [2]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
from IPython.display import display

#Libreria para plotear
%pip install --upgrade plotly
import plotly.express as px

# Librerias utiles
from sklearn.preprocessing import StandardScaler

Note: you may need to restart the kernel to use updated packages.


# 1. Rendimiento en Estudiantes 📚

Para este laboratorio deberán trabajar con el dataset `students_grades.json` el cual contiene información sobre el rendimiento de alumnos en distintos ramos.
El objetivo principal será el de realizar un trabajo de limpieza de prepración y exploración de datos utilizando exclusivamente `DataFrames` y sus métodos asociados.




In [3]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = 'Dirección donde tiene los archivos en el Drive'
except:
    print('Ignorando conexión drive-colab')

Ignorando conexión drive-colab


## 1. Carga e Inspección de Datos [1.0 Puntos]


<p align="center">
  <img src="https://i.imgur.com/44knnIf.png" width="500" height="500">
</p>

*Don Caguayo*, rector de la Universidad de la Cachaña, les comenta que actualmente la Universidad no tiene una subdirección de análisis de datos, pero que tiene un gran interés en formar una. Es por esto, que se contacta con ud. y su equipo para iniciar de forma exploratoria diversas tareas de análisis y consolidación de datos respectivos al rendimiento de sus alumnos.

Previo a su primera reunión con el rector, el rector le envía una copia de la base de datos, la cuál a usted y su equipo le encantaría examinar. Para esto, la primera tarea que debe realizar es abrir y cargar el dataset.


> **Notar que es sumamente importante el formato que posee cada columna** para su manipulación y que posiblemente ustedes deberán especificar algunos tipos de datos a mano. Ver la documentación del método `read_json` para mayor información.

A continuación se muestra un ejemplo de los dos primeros registros del dataset:

```json
[
  {
    "names": "Rita Courtney",
    "gender": "female",
    "race/ethnicity": "group B",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 37,
    "reading score": 46,
    "writing score": 46
  },
  {
    "names": "Madeline Fuller",
    "gender": "female",
    "race/ethnicity": "group C",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 124,
    "reading score": 142,
    "writing score": 142
  }
]
```





In [4]:
df_grades = pd.read_json('students_grades.json')


<p align="center">
  <img src="https://i.giphy.com/media/4FQMuOKR6zQRO/giphy.webp">
</p>


Posteriormente, el rector les comenta que la sección de TI posee una base de datos con más de **1000** filas, en donde cada fila representa un alumno matriculado. Sin embargo, cree fuertemente que existen errores en esta: solo tiene registro de **1000** matriculas. Por este motivo, especula que se han realizado malas prácticas al registrar dichos datos.

En consecuencia, el rector les solicita a ustedes como tarea inicial que exploren la base de datos y encuentre posibles falencias dentro de sus registros. Para esto, usted debe estar al tanto que las notas están en una escala que van del **0 al 100** (porcentaje de logro) y **no es posible que un alumno haya dado más de una vez el mismo examen**. Cualquier registro que posea un valor diferente a esta escala, **debería ser eliminado de la base de datos**.


> **Instrucciones**

**Cree una función** que tome un `DataFrame` y que le permita revisar si las diferentes variables que lo componen son coherentes y si es posible observar anomalías en estas.

**To-Do**
1. [ ] Imprimir un mensaje con las dimensiones del Dataframe.
2. [ ] Imprimir un mensaje con los nombres de las columnas que conforman el DataFrame.
3. [ ] Mostrar los primeros 5, los últimos 5 elementos del DataFrame y realice un muestreo de 5 elementos de forma aleatoria.
4. [ ] Mostrar una descripción rápida de las variables numéricas del DataFrame (utilice un método visto en clases).
5. [ ] Imprimir un mensaje con la cantidad de valores nulos (investigue el método `isna()` y `sum()` de los DataFrames).
6. [ ] Imprimir un mensaje que detalle la cantidad de valores unicos (investigue `nunique()`) .
7. [ ] Imprimir un mensaje con el total de filas que contengan nombres duplicados (investigue `duplicated()` y su parámetro `subset`).
8. [ ] Mostrar las filas que contengan nombres sean duplicados (investigue `duplicated()` y su parámetro `subset`).

Finalmente, responda las siguientes preguntas (comente esto en la celda ubicada debajo de **justificación**.):

9. [ ] ¿Qué anomalías es posible observar de los datos que componen el DataFrame? ¿Existen valores nulos o duplicados?


10. [ ] Su compañero ChatGPT le provee un código para que pueda entender los resultados de las pruebas desde un gráfico. ¿Qué puede interpretar de la distribución de las notas en el boxplot?

> **Nota:** Para imprmir utilice `print(...)`. Para mostrar `DataFrames`, utilice `display(...)` ya que este permite mostrar estos elementos como tablas.

> **Nota 2:** Utilice la documentación de [pandas](https://pandas.pydata.org/docs/index.html) como también StackOverflow para resolver cualquier problema o inconveniente.

> **Nota 3:** Si no ha trabajado con boxplot antes, el siguiente link puede serle de mucha ayuda: [boxplot](https://medium.com/dayem-siddiqui/understanding-and-interpreting-box-plots-d07aab9d1b6c)

**Respuesta:**

In [5]:
def exploratory_data_analysis(dataframe):
    row, column = dataframe.shape
    print(f'1. La dimensión es {row, column}')

    column = dataframe.columns
    print(f'2. Los nombres de las columnas son: {column}')

    head = dataframe.head(5)
    print('3. Los primeros 5 elementos son:')
    display(head)

    tail = dataframe.tail(5)
    print('3. Los últimos 5 elementos son:')
    display(tail)

    aleatory = dataframe.sample(5)
    print('3. Muestras aleatorias:')
    display(aleatory)

    description = dataframe.describe()
    print('4. Descripción rápida de las variables')
    display(description)

    nulls = dataframe.isna().sum()
    print('5. Cantidad de valores nulos')
    display(nulls)

    uniques = dataframe.nunique()
    print('6. Cantidad de valores únicos')
    display(uniques)

    total_duplicados = dataframe.duplicated(subset='names').sum()
    print(f'7. Cantidad de nombres duplicados: {total_duplicados}')

    nombres_duplicados = df_grades[df_grades.duplicated(subset=['names'])]
    print('8. Filas que contienen nombres duplicados:')
    display(nombres_duplicados)


In [6]:
# Código del amigo ChatGPT (listo para usar)
fig = px.box(
    df_grades.melt(
        value_vars=["math score", "reading score", "writing score"],
        var_name="subject",
        value_name="score",
    ),
    y="score",
    color="subject",
)
fig.show()

In [7]:
exploratory_data_analysis(df_grades)

1. La dimensión es (1400, 9)
2. Los nombres de las columnas son: Index(['names', 'gender', 'race/ethnicity', 'parental level of education',
       'lunch', 'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')
3. Los primeros 5 elementos son:


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course math score reading score writing score  
0  standard                    none         37            46            46  
1  standard                    none        124           142           142  
2  standard                    none        156           196           193  
3  standard               completed         80            78            81  
4  standard                    none        131           166           166

3. Los últimos 5 elementos son:


names  gender race/ethnicity parental level of education  \
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
1395      standard                    none        193           142   
1396  free/reduced               completed         66            74   
1397      standard               completed         54            61   
1398  free/reduced                    none        193           196   
1399      standard               completed         68            80   

     writing score  
1395           193  
1396            81  
1397            62  
1398           193  
1399            76

3. Muestras aleatorias:


names  gender race/ethnicity parental level of education  \
1390   Gloria Bradford  female        group B                 high school   
547      Oliver Seldon    male        group B                 high school   
931       Jeff Pickett    male        group E          associate's degree   
684   Marlene Sullivan  female        group D             master's degree   
1198    Susan Lairmore  female        group C          associate's degree   

             lunch test preparation course math score reading score  \
1390      standard                    none        124            64   
547       standard               completed         76            62   
931       standard                    none        193           193   
684       standard               completed         77            82   
1198  free/reduced                    none        124           193   

     writing score  
1390           193  
547             60  
931            193  
684             91  
1198           193

4. Descripción rápida de las variables


names  gender race/ethnicity parental level of education  \
count             1400    1400           1400                        1400   
unique            1153       2              5                           6   
top     Robert Frazier  female        group C                some college   
freq                 3     723            452                         334   

           lunch test preparation course  math score  reading score  \
count       1400                    1400        1400           1400   
unique         2                       2          97             89   
top     standard                    none         193            193   
freq         904                     898         120            117   

        writing score  
count            1400  
unique             93  
top               193  
freq              133

5. Cantidad de valores nulos


names                          0
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

6. Cantidad de valores únicos


names                          1153
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       97
reading score                    89
writing score                    93
dtype: int64

7. Cantidad de nombres duplicados: 247
8. Filas que contienen nombres duplicados:


names  gender race/ethnicity parental level of education  \
12         Rita Courtney  female        group B            some high school   
81    Elizabeth Quintero  female        group B                some college   
160    Michelle Gonzales  female        group B            some high school   
170        Carla Mendoza  female        group D          associate's degree   
176       Darlene Parker  female        group C          associate's degree   
...                  ...     ...            ...                         ...   
1385      James Crawford    male        group D             master's degree   
1390     Gloria Bradford  female        group B                 high school   
1393         Adam Romero    male        group C                 high school   
1395    Alexander Dillon    male        group D          associate's degree   
1398        Joseph Mccoy    male        group D                some college   

             lunch test preparation course math score reading score  \
12        standard                    none        193           193   
81        standard                    none         82            85   
160       standard               completed         60            70   
170   free/reduced                    none         43            60   
176       standard                    none         54            61   
...            ...                     ...        ...           ...   
1385      standard                    none        193           147   
1390      standard                    none        124            64   
1393      standard               completed        124           193   
1395      standard                    none        193           142   
1398  free/reduced                    none        193           196   

     writing score  
12             193  
81              87  
160             74  
170             58  
176             58  
...            ...  
1385           131  
1390           193  
1393            49  
1395           193  
1398           193  

[247 rows x 9 columns]

**Justificación:**

9. A partir de lo obtenido anteriormente, se observa que, aunque no hay valores nulos, existen 247 nombres duplicados, siendo Robert Frazier el más repetido. Esta cantidad de duplicaciones explica por qué el director del colegio cuenta con más de 1000 registros, a pesar de que el total de alumnos es de solo 1000.

 
10. Del boxplot se puede notar que las distribuciones de notas en las pruebas de lectura y escritura son bastante similares, con escritura mostrando una mediana ligeramente inferior. Por otro lado, en matemáticas se observan los peores resultados, siendo la mediana y la nota mínima más bajas en comparación con las demás asignaturas. 

## 2 Limpieza de Datos [1.0 puntos]




<p align="center">
  <img src="https://blogdatlas.files.wordpress.com/2020/07/datlas_mx_blog_limpieza_bds_data-scrubbing-service.gif">
</p>

En base a lo realizado anteriormente, realice una limpieza de los dataset. Asuma los supuestos señalados en la pregunta anterior para limpiar el `DataFrame` y **cree una función** que le permita limpiar los datos. Habiendo limpiado el `DataFrame`, compruebe nuevamente los datos con la función creada en la sección 1.1.

**To-Do:**
- [ ] Eliminar registros con valores nulos.
- [ ] Eliminar valores fuera del rango válido.
- [ ] Eliminar registros con nombres duplicados
- [ ] Comprobar que no existen anomalías en los datos a traves de la función de la sección 1.1.

**Hints**

- Puede ser útil los métodos ```.drop_duplicates()``` y ```.dropna()``` como también utilizar queries con máscaras booleanas.
- Cuidado con el orden en la limpieza de datos. ¿Que se debe eliminar primero, los nombres duplicados o los registros fuera del rango posible?


> Utilice la función creada en el punto 1 para verificar sus resultados.

**Respuesta:**

In [8]:
def clean_data(dataframe):
    """ 
    Limpia el dataframe de valores nulos, fuera de rango y duplicados.

    Args:
        dataframe: DataFrame a limpiar

    Returns:
        DataFrame limpio
    """
    # Se reemplazan valores vacios por NaN
    dataframe = dataframe.replace('', pd.NA)

    # Se eliminan registros con valores nulos
    dataframe = dataframe.dropna()

    # Se eliminan valores fuera del rango valido (0-100) para todas las columnas de notas
    dataframe = dataframe[(dataframe['math score'] >= 0) & (dataframe['math score'] <= 100)]
    dataframe = dataframe[(dataframe['reading score'] >= 0) & (dataframe['reading score'] <= 100)]
    dataframe = dataframe[(dataframe['writing score'] >= 0) & (dataframe['writing score'] <= 100)]

    # Se eliminan valores duplicados (un alumno dando el mismo examen dos veces)
    dataframe = dataframe.drop_duplicates(subset=['names'])

    return dataframe

In [9]:
#Pruebe su código aquí
df_grades = clean_data(df_grades)
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
0         standard                    none         37            46   
3         standard               completed         80            78   
5         standard                    none         73            66   
6         standard               completed         81            75   
7         standard               completed         94            86   
...            ...                     ...        ...           ...   
1392      standard                    none         69            75   
1394  free/reduced               completed         23            44   
1396  free/reduced               completed         66            74   
1397      standard               completed         54            61   
1399      standard               completed         68            80   

     writing score  
0               46  
3               81  
5               63  
6               76  
7               87  
...            ...  
1392            71  
1394            36  
1396            81  
1397            62  
1399            76  

[1000 rows x 9 columns]

In [10]:
exploratory_data_analysis(df_grades)

1. La dimensión es (1000, 9)
2. Los nombres de las columnas son: Index(['names', 'gender', 'race/ethnicity', 'parental level of education',
       'lunch', 'test preparation course', 'math score', 'reading score',
       'writing score'],
      dtype='object')
3. Los primeros 5 elementos son:


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course math score reading score writing score  
0  standard                    none         37            46            46  
3  standard               completed         80            78            81  
5  standard                    none         73            66            63  
6  standard               completed         81            75            76  
7  standard               completed         94            86            87

3. Los últimos 5 elementos son:


names  gender race/ethnicity parental level of education  \
1392   Richard Young    male        group D                 high school   
1394   Wanda Russell  female        group B                 high school   
1396  Marina Zeigler  female        group C           bachelor's degree   
1397   Laurie Carter  female        group B            some high school   
1399    Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
1392      standard                    none         69            75   
1394  free/reduced               completed         23            44   
1396  free/reduced               completed         66            74   
1397      standard               completed         54            61   
1399      standard               completed         68            80   

     writing score  
1392            71  
1394            36  
1396            81  
1397            62  
1399            76

3. Muestras aleatorias:


names gender race/ethnicity parental level of education  \
771   Alex Estimable   male        group C                 high school   
1230     Larry Braun   male        group C            some high school   
735   Larry Matthews   male        group B          associate's degree   
1081   James Aguirre   male        group E                 high school   
1163  Nathan Mewborn   male        group D             master's degree   

             lunch test preparation course math score reading score  \
771       standard               completed         58            52   
1230  free/reduced               completed         53            37   
735   free/reduced                    none         44            41   
1081      standard               completed         81            80   
1163      standard                    none         73            70   

     writing score  
771             54  
1230            40  
735             38  
1081            76  
1163            75

4. Descripción rápida de las variables


names  gender race/ethnicity parental level of education  \
count            1000    1000           1000                        1000   
unique           1000       2              5                           6   
top     Rita Courtney  female        group C                some college   
freq                1     518            319                         226   

           lunch test preparation course  math score  reading score  \
count       1000                    1000        1000           1000   
unique         2                       2          81             72   
top     standard                    none          65             72   
freq         645                     642          36             34   

        writing score  
count            1000  
unique             77  
top                74  
freq               35

5. Cantidad de valores nulos


names                          0
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

6. Cantidad de valores únicos


names                          1000
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       81
reading score                    72
writing score                    77
dtype: int64

7. Cantidad de nombres duplicados: 0
8. Filas que contienen nombres duplicados:


Empty DataFrame
Columns: [names, gender, race/ethnicity, parental level of education, lunch, test preparation course, math score, reading score, writing score]
Index: []

**Justificación:**

Para una correcta limpieza de datos, se debe eliminar primero los valores fuera de rango ya que se permite limpiar los datos de manera más precisa. Por ejemplo, si se tienen dos filas duplicadas pero una de ellas contiene valores fuera de rango (o inválidos), eliminar los duplicados primero podría implicar perder una fila con información válida. Por lo que de cierta manera al eliminar primero los valores fuera de rango, se asegura conservar los datos correctos antes de manejar los duplicados.

## 3. Trabajando con Datos [2.0 puntos]

<p align="center">
  <img src="https://media3.giphy.com/media/YQitE4YNQNahy/giphy-downsized-large.gif">
</p>

### 3.1 Transformación de notas a "Nota Chilena" [0.3 Puntos]

En una de las reuniones con Don Caguayo este les comenta su intención de pasar todas sus notas a la "Escala Chilena". Para esto les facilita la siguiente ecuación:

$$Nota \, Chilena= \% \, de \, logro*0.06 + 1$$

Utilizando la ecuación entregada, cree una función que le permita pasar a la escala Chilena las notas del dataframe y entregue las notas redondeadas al segundo decimal.

**To-Do**

- [ ] Tranforme los valores a la escala chilena.

**Respuesta:**

In [11]:
def nota_chilena(dataframe_in, columna):
    """
    Convierte las notas de una columna a escala chilena.
    
    Args:
        dataframe_in: DataFrame con las notas a convertir
        columna: Nombre de la columna con las notas a convertir
    
    Returns:
        DataFrame con las notas convertidas
    """
    # Se copia dataframe
    df = dataframe_in.copy()
    
    # Se convierten las notas a la escala chilena
    df[columna] = df[columna].apply(lambda x: (x*0.06) + 1)
    
    return df

In [12]:
#Muestre sus resultados aquí
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
0         standard                    none         37            46   
3         standard               completed         80            78   
5         standard                    none         73            66   
6         standard               completed         81            75   
7         standard               completed         94            86   
...            ...                     ...        ...           ...   
1392      standard                    none         69            75   
1394  free/reduced               completed         23            44   
1396  free/reduced               completed         66            74   
1397      standard               completed         54            61   
1399      standard               completed         68            80   

     writing score  
0               46  
3               81  
5               63  
6               76  
7               87  
...            ...  
1392            71  
1394            36  
1396            81  
1397            62  
1399            76  

[1000 rows x 9 columns]

In [13]:
nota_chilena(df_grades, ['math score', 'reading score', 'writing score'])

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
0         standard                    none       3.22          3.76   
3         standard               completed        5.8          5.68   
5         standard                    none       5.38          4.96   
6         standard               completed       5.86           5.5   
7         standard               completed       6.64          6.16   
...            ...                     ...        ...           ...   
1392      standard                    none       5.14           5.5   
1394  free/reduced               completed       2.38          3.64   
1396  free/reduced               completed       4.96          5.44   
1397      standard               completed       4.24          4.66   
1399      standard               completed       5.08           5.8   

     writing score  
0             3.76  
3             5.86  
5             4.78  
6             5.56  
7             6.22  
...            ...  
1392          5.26  
1394          3.16  
1396          5.86  
1397          4.72  
1399          5.56  

[1000 rows x 9 columns]

### 3.2 Obtención de promedio de alumnos [0.3 Puntos]

Con todas las notas adjuntadas, la prestigiosa Universidad De la Cachaña necesita obtener el promedio general de cada uno de los alumnos. Para esto se le solicita que calcule el promedio con cada una de las asignaturas y el promedio se anexe como una nueva columna al final del `DataFrame` original. Cabe señalar que las notas deben ser redondeadas al segundo decimal.

**To-Do:**
- [ ] Crear función que nos permite obtener el promedio de cada uno de los alumnos y anexar esta información en el `DataFrame` original.


**Respuesta:**

In [14]:
def alumnos_promedio(dataframe_in):
    """
    Calcula el promedio de las notas de los alumnos y agrega una columna con el promedio.
    Args:
        dataframe_in: DataFrame
    Returns:
        DataFrame con una columna adicional llamada 'mean_test'
    """
    # Copiamos dataframe
    df = dataframe_in.copy()
    
    # Se calcula el promedio de las notas
    df['GPA'] = df[['math score', 'reading score', 'writing score']].mean(axis=1).round(2)
    
    return df

In [15]:
#Muestre sus resultados aquí
df_grades = alumnos_promedio(df_grades)
df_grades.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course math score reading score writing score  \
0  standard                    none         37            46            46   
3  standard               completed         80            78            81   
5  standard                    none         73            66            63   
6  standard               completed         81            75            76   
7  standard               completed         94            86            87   

         GPA  
0       43.0  
3  79.666667  
5  67.333333  
6  77.333333  
7       89.0

### 3.3 Reprobados y Destacados [0.9 Puntos, 0.3 cada DataFrame]

Contentos con su desempeño, la Universidad les solicita que obtengan los alumnos quienes han reprobado el año y los que han tenido desempeño de destacado. Para esto se considerará que un alumno ha reprobado el año cuando el promedio de notas es inferior a 4.0 en la Escala Chilena, y se considerará destacado si posee notas superiores a 5.8 en todas las asignaturas (o sea notas superiores a 5.8 en `math score`, `reading score` y `writing score`).

Cree una función que permita obtener tres Dataframes:
- El primero deberá poseer solamente los alumnos que han reprobado en la prestigiosa universidad.
- El segundo debe entregar un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados.
- El tercero será un Data Frame con los alumnos destacados ordenados de mayor a menor.

**To-Do:**

- [ ] Obtener DataFrame con los alumnos que reprobaron el año.
- [ ] Obtener DataFrame similar al original, pero marcados con una "R" el atributo GPA de los alumnos reprobados.
- [ ] Obtener DataFrame con los alumnos destacados ordenados de mayor a menor.

**Respuesta:**

In [16]:
def reprobados_destacados(dataframe_in):
    # Se obtiene dataframe con alumnos reprobados (promedio de notas menor a 4)
    df_reprobados = dataframe_in[(dataframe_in['GPA'] < 40)]

    # Se obtiene dataframe similar al original, pero marcados con R el atributo GPA de los reprobados
    df_rwr = dataframe_in.copy()
    df_rwr['GPA'] = df_rwr['GPA'].apply(lambda x: 'R' if x < 4.0 else x)

    # Se obtiene dataframe con alumnos destacados (nota superior a 58 en todas las asignaturas) y ordenados de mayor a menor
    df_dest = dataframe_in.copy()
    df_dest = df_dest[(df_dest['math score'] > 58) & (df_dest['reading score'] > 58) & (df_dest['writing score'] > 58)]
    df_dest = df_dest.sort_values(by='GPA', ascending=False)

    return df_reprobados, df_rwr, df_dest

In [17]:
# Muestre sus resultados aquí
dataframe1, dataframe2, dataframe3 = reprobados_destacados(df_grades)

In [18]:
dataframe1.head()

names  gender race/ethnicity parental level of education  \
55     Rodger Talavera    male        group B                 high school   
66     Anthony Malcomb    male        group C                some college   
196    Patricia Briney  female        group B                some college   
311     Gerald Brunner    male        group D                 high school   
312  Barbara Henriques  female        group C                some college   

            lunch test preparation course math score reading score  \
55   free/reduced                    none         36            29   
66   free/reduced                    none         35            28   
196      standard                    none         19            38   
311      standard                    none         46            34   
312  free/reduced                    none         22            39   

    writing score        GPA  
55             27  30.666667  
66             27       30.0  
196            32  29.666667  
311            36  38.666667  
312            33  31.333333

In [19]:
dataframe2.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course math score reading score writing score  \
0  standard                    none         37            46            46   
3  standard               completed         80            78            81   
5  standard                    none         73            66            63   
6  standard               completed         81            75            76   
7  standard               completed         94            86            87   

         GPA  
0  43.000000  
3  79.666667  
5  67.333333  
6  77.333333  
7  89.000000

In [20]:
dataframe3.head()

names  gender race/ethnicity parental level of education  \
677       Richard Pryor    male        group E           bachelor's degree   
384   Elizabeth Beasley  female        group E           bachelor's degree   
1352      Sandra Pompey  female        group E          associate's degree   
475           Emma Gray  female        group E           bachelor's degree   
1276     Angela Cressey  female        group D                some college   

         lunch test preparation course math score reading score writing score  \
677   standard               completed        100           100           100   
384   standard                    none        100           100           100   
1352  standard                    none        100           100           100   
475   standard               completed         99           100           100   
1276  standard                    none         98           100            99   

            GPA  
677       100.0  
384       100.0  
1352      100.0  
475   99.666667  
1276       99.0

### 3.4 Quintiles de Rendimiento [0.5 Puntos]

El rector, intrigado por saber si existe una relación entre el nivel educacional de los padres y las notas de sus hijos les plantea estudiar dicha relación. Para esto, les solicita que desde el `DataFrame` que posee a todos los alumnos (`DataFrame` obtenido tras la limpieza), generen una nueva columna que clasifique en quintiles de rendimiento para sus alumnos: `['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto']`. (investigar `qcut()`)

Hecho esto, grafique a través de un scatter plot el nivel educacional de los padres v/s el promedio de los alumnos, utilice los cuantiles generados anteriormente para mostrar en que cuantil se encuentra cada uno de los datos.

Realizado el grafico, responda la siguiente pregunta: **¿Es posible visualizar alguna tendencia en el gráfico?.**



**Respuesta:**

In [21]:
"""
NOTA: La variable del dataframe que contenga los quintiles debe llamarse
'GPA cuantiles' para que el gráfico funcione.
"""
# Se obtienen los quintiles de la columna GPA
df_grades['GPA cuantiles'] = pd.qcut(df_grades['GPA'], 5, labels= ['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto'])

In [22]:
# Código de gráfico del amigo ChatGPT
fig = px.scatter(
    df_grades.sort_values(by="GPA"),
    x="GPA",
    y="parental level of education",
    hover_data=["gender"],
    color="GPA cuantiles",
    title="GPA por Nivel de educación de los padres",
    color_discrete_sequence=["tomato", "orange", "yellow", "greenyellow", "green"]
)

fig

Se puede observar que cuando los padres posee un master's degree, los promedios de los alumnos se distribuyen más hacia la derecha, lo que indica que son más altos. Al mismo tiempo se observa que en este caso no hay tantos promedios ubicados en el nivel bajisimo de GPA. Por otro lado, se aprecia que cuando los padres poseen un nivel de educación como: associate's degree, some college y high school, se presenta un comportamiento similar en cuanto a la distribución de GPA de los alumnos.

## 4. Operaciones con tablas [2.0 puntos]

<p align="center">
  <img src="https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYWx5dTBkMGhhOTJmeGQzbDV5M21rNTN1NWVwMm1qbHd3cjh3Nmk4NSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/13HBDT4QSTpveU/giphy.gif">
</p>



### 4.1 Carga de datos y concatenar dataset [0.2 Puntos]

Ya finalizado en análisis inicial, ud. y su equipo le entregaron a *Don Caguayo* (rector de la Universidad de la Cachaña) tanto los resultados del análisis como también la base de datos limpia y lista para ser almacenada. Dada la ingente cantidad de los datos, el equipo de TI de la universidad resolvió separar el dataset en dos bases de datos distintas (lo que según argumentan ellos, permitiría hacer agregaciones de forma más eficiente).

Gracias a la excelente labor de ud. y su equipo en el análisis previo, el rector le solicita continuar el trabajo con una nueva batería de análisis. Por este motivo, la sección de TI les entrega nuevamente los datos. Sin embargo, argumentan que dada una escazes de personal, solo le entregarán *dumps* (copias) de cada base de datos y su equipo deberá unir las bases de datos. Los datos se encuentran en los siguiente archivos `.json`: `students_grades_1.json` y `students_grades_2.json`.


Por ende, ud. y su equipo deciden que la primera tarea se centrará en cargar estos datos y unirlos.

**No se preocupe por la limpieza ni transformar el tipo de datos de las columnas, ni tampoco transformar a notas chilenas**, recuerde que anteriormente ya se encargo de este tema.


Hint: investigue el método `concat`

**Respuesta:**

In [26]:
df_grades_1= pd.read_json('students_grades_1.json')
df_grades_2= pd.read_json('students_grades_2.json')

df_grades= pd.concat([df_grades_1, df_grades_2])
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1     Charles Linstrom    male        group A           bachelor's degree   
2          Brian Young    male        group C            some high school   
3       Howard Jimenez    male        group E            some high school   
4         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
998      Richard Young    male        group D                 high school   
999      Wanda Russell  female        group B                 high school   
1000    Marina Zeigler  female        group C           bachelor's degree   
1001     Laurie Carter  female        group B            some high school   
1002      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
1         standard               completed        5.80           5.68   
2         standard                    none        5.38           4.96   
3         standard               completed        5.86           5.50   
4         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
998       standard                    none        5.14           5.50   
999   free/reduced               completed        2.38           3.64   
1000  free/reduced               completed        4.96           5.44   
1001      standard               completed        4.24           4.66   
1002      standard               completed        5.08           5.80   

      writing score   GPA GPA cuantiles  
0              3.76  3.58      Bajísimo  
1              5.86  5.78          Alto  
2              4.78  5.04         Medio  
3              5.56  5.64          Alto  
4              6.22  6.34      Muy alto  
...             ...   ...           ...  
998            5.26  5.30         Medio  
999            3.16  3.06      Bajísimo  
1000           5.86  5.42          Alto  
1001           4.72  4.54          Bajo  
1002           5.56  5.48          Alto  

[1000 rows x 11 columns]

### 4.2 Análisis por Nivel Educacional Etnia de los Padres [0.8 Puntos]

El rector, basado en su experiencia, cree fuertemente que el nivel educacional y la etnia de los padres influyen en las notas que obtienen sus hijos.
Como científicos de datos, ud. y su equipo creen que deben encontrar evidencia para confirmar o refutar la hipótesis del rector.

Para esto, deciden generar dos análisis: una tabla de resumen por una parte y gráficos de caja por otro.


**Tabla de Resumen [0.5 Puntos]**


Para generar la tabla de resumen:

- [ ] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
- [ ] Hacer una simplificación a través de un mapeo (investigar el método `map()`) de la variable `parental level of education`  según la siguiente conversión:

        some high school -> school
        some college -> school
        high school -> school
        bachelor's degree -> college
        associate's degree -> college
        master's degree -> postgraduate

  Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas.
- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
- [ ] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'.

Utilizar la tabla de resultados esperados como guía para desarrollar este punto.


**Análisis [0.3 Puntos]**

Finalmente, responda las siguientes preguntas sobre el gráfico de caja que viene implementado a continuación. En este se muestra el GPA con respecto al nivel educacional, donde la variable de color es la etnicidad.

- [ ]  ¿Hay alguna diferencia entre los grupos graficados tanto para el nivel educacional de los padres como también para la etnicidad?


**Respuesta:**

In [34]:
# Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
df_grades['GPA'] = df_grades[['math score', 'reading score', 'writing score']].mean(axis=1).round(2)

# Se realiza una simplificación a través de un mapeo de la columna 'parental level of education'.

# Se define el mapeo
mapeo = {'some high school': 'school',
    'some college': 'school',
    'high school': 'school',
    "bachelor's degree": 'college',
    "associate's degree": 'college',
    "master's degree": 'postgraduate'}

# Se agrega la columna respectiva y se hace el mapeo
df_grades['simple parental level of education'] = df_grades['parental level of education'].map(mapeo)

# Agregar según `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas.
df_grades_grouped = df_grades.groupby(['race/ethnicity', 'simple parental level of education'])[['math score', 'reading score', 'writing score']].mean().round(2)

# Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener conteo de alumnos por grupo
df_grades_grouped['count'] = df_grades.groupby(['race/ethnicity', 'simple parental level of education']).size()

# Obtener el porcentaje de alumnos con respecto al total. 
df_grades_grouped['percentage'] = (df_grades_grouped['count'] / df_grades_grouped['count'].sum() * 100).apply(lambda x: f'{x}%')

In [35]:
df_grades_grouped

math score  reading score  \
race/ethnicity simple parental level of education                              
group A        college                                   4.83           5.05   
               postgraduate                              4.46           4.88   
               school                                    4.65           4.81   
group B        college                                   5.03           5.24   
               postgraduate                              5.03           5.81   
               school                                    4.69           4.90   
group C        college                                   5.03           5.36   
               postgraduate                              5.02           5.23   
               school                                    4.74           5.00   
group D        college                                   5.06           5.22   
               postgraduate                              5.35           5.63   
               school                                    4.99           5.13   
group E        college                                   5.53           5.45   
               postgraduate                              5.48           5.93   
               school                                    5.35           5.27   

                                                   writing score  count  \
race/ethnicity simple parental level of education                         
group A        college                                      4.95     26   
               postgraduate                                 5.06      3   
               school                                       4.66     60   
group B        college                                      5.16     61   
               postgraduate                                 5.63      6   
               school                                       4.79    123   
group C        college                                      5.33    118   
               postgraduate                                 5.17     19   
               school                                       4.89    182   
group D        college                                      5.24     78   
               postgraduate                                 5.78     23   
               school                                       5.11    161   
group E        college                                      5.43     57   
               postgraduate                                 5.83      8   
               school                                       5.11     75   

                                                            percentage  
race/ethnicity simple parental level of education                       
group A        college                                            2.6%  
               postgraduate                                       0.3%  
               school                                             6.0%  
group B        college                                            6.1%  
               postgraduate                                       0.6%  
               school                                            12.3%  
group C        college                             11.799999999999999%  
               postgraduate                                       1.9%  
               school                                            18.2%  
group D        college                                            7.8%  
               postgraduate                                       2.3%  
               school                                            16.1%  
group E        college                                            5.7%  
               postgraduate                                       0.8%  
               school                                             7.5%

In [36]:
fig = px.box(
    df_grades,
    x="simple parental level of education",
    y="GPA",
    color="race/ethnicity",
    hover_data=["GPA"],
)
fig.show()

Se puede ver que hay diferencias notorias tanto para el nivel educacional como para la etnicidad, en particular, se puede ver que mientras mayor es el nivel educacional, los promedios de GPA tienden a subir. A la vez, se puede apreciar que para ciertos grupos etnicos hay una mayor tendencia a GPA's más altos, como es el caso del grupo B y el grupo E, los cuales tienden a mayores GPA que los demás grupos, especialmente en el caso en que los padres tienen un posgrado. Por lo tanto, se concluye que existe una diferencia entre grupos causado por el nivel educacional y la etnicidad de los padres.

**Resultados Esperados**

Esta tabla ilustra la estructura del dataframe a conseguir. Los valores son referenciales, no necesariamente son los esperados.

|    | race/ethnicity   | simple parental level of education   |   math score |   reading score |   writing score |   GPA |   count | percentage   |
|---:|:-----------------|:-------------------------------------|-------------:|----------------:|----------------:|------:|--------:|:-------------|
|  0 | group A          | college                              |         4.83 |            5.05 |            4.95 |  4.94 |      26 | 2.6 %        |
|  1 |                  | postgraduate                         |         4.46 |            4.88 |            5.06 |  4.80 |       3 | 0.3 %        |
|  2 |                  | school                               |         4.65 |            4.81 |            4.66 |  4.71 |      60 | 6.0 %        |
|  3 | group B          | college                              |         5.03 |            5.24 |            5.16 |  5.14 |      61 | 6.1 %        |
|  4 |                  | postgraduate                         |         5.03 |            5.81 |            5.63 |  5.49 |       6 | 0.6 %        |
|  5 |                  | school                               |         4.69 |            4.90 |            4.79 |  4.79 |     123 | 12.3 %       |
|  6 | group C          | college                              |         5.03 |            5.36 |            5.33 |  5.24 |     118 | 11.8 %       |
|  7 |                  | postgraduate                         |         5.02 |            5.23 |            5.17 |  5.14 |      19 | 1.9 %        |
|  8 |                  | school                               |         4.74 |            5.00 |            4.89 |  4.88 |     182 | 18.2 %       |
|  9 | group D          | college                              |         5.06 |            5.22 |            5.24 |  5.17 |      78 | 7.8 %        |
| 10 |                  | postgraduate                         |         5.35 |            5.63 |            5.78 |  5.59 |      23 | 2.3 %        |
| 11 |                  | school                               |         4.99 |            5.13 |            5.11 |  5.08 |     161 | 16.1 %       |
| 12 | group E          | college                              |         5.53 |            5.45 |            5.43 |  5.47 |      57 | 5.7 %        |
| 13 |                  | postgraduate                         |         5.48 |            5.93 |            5.83 |  5.74 |       8 | 0.8 %        |
| 14 |                  | school                               |         5.35 |            5.27 |            5.11 |  5.25 |      75 | 7.5 %        |


### 4.3 Joins, joins y más joins [1 punto]

Mientras le notificaba por videollamada los resultados de sus descubrimientos a Don Caguayo, un exaltado practicante del area de TI entra a la reunión y les informa que ha encontrado una nueva base de datos que cuenta con las notas de dos asignaturas (en escala chilena): historia y ciencias.
Para más remate, antes de huir, el practicante les cuenta que este dataframe lamentablemente contiene registros de alumnos que ya no forman parte de la universidad, por lo que no deben ser parte de este análisis.


El rector (evidentemente molesto por la situación) les ruega incluir estos datos (vaciados en el archivo other_grades.csv) al estudio original(`students_grades.csv`).


Para esto, carge el archivo `other_grades.csv` y busque la forma de unir ambos DataFrames, de tal manera que las columnas de `history score` y `science score` se anexen al final del DataFrame original. **NO LIMPIE LOS DATOS**, si no que explore los distintos tipos de merge para encontrar el mas situable para su situación (y así evitar buscar duplicados o **alumnos que no pertenecen a la universidad**).

**To-Do**


- [ ] Cargar el `other_grades.csv`
- [ ] Unir `df_grades` con `other_grades.csv` usando outer join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando left join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando right join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando inner join y explique el resultado.
- [ ] Defina cuál join es el que utilizará para generar el nuevo DataFrame.

> **Hint**: Puede explicar los resultados del merge a través de la cantidad de filas resultantes y los valores que estas contienen.

**Respuesta:**

In [ ]:
# Inserte su respuesta aquí
other_grades = pd.read_csv('other_grades.csv')

df_grades_outer = pd.merge(df_grades, other_grades, on='names', how='outer')
df_grades_left = pd.merge(df_grades, other_grades, on='names', how='left')
df_grades_right = pd.merge(df_grades, other_grades, on='names', how='right')
df_grades_inner = pd.merge(df_grades, other_grades, on='names', how='inner')

display(df_grades_outer)

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1     Charles Linstrom    male        group A           bachelor's degree   
2          Brian Young    male        group C            some high school   
3       Howard Jimenez    male        group E            some high school   
4         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1145     Carson Brooks     NaN            NaN                         NaN   
1146         Ruby Reed     NaN            NaN                         NaN   
1147    Zachary Torres     NaN            NaN                         NaN   
1148       Eliana Gray     NaN            NaN                         NaN   
1149       Gavin James     NaN            NaN                         NaN   

         lunch test preparation course math score reading score writing score  \
0     standard                    none         37            46            46   
1     standard               completed         80            78            81   
2     standard                    none         73            66            63   
3     standard               completed         81            75            76   
4     standard               completed         94            86            87   
...        ...                     ...        ...           ...           ...   
1145       NaN                     NaN        NaN           NaN           NaN   
1146       NaN                     NaN        NaN           NaN           NaN   
1147       NaN                     NaN        NaN           NaN           NaN   
1148       NaN                     NaN        NaN           NaN           NaN   
1149       NaN                     NaN        NaN           NaN           NaN   

        GPA GPA cuantiles  science score  history score  
0     43.00      Bajísimo           2.50           3.80  
1     79.67          Alto           4.36           5.20  
2     67.33         Medio           4.48           3.80  
3     77.33          Alto           5.32           5.00  
4     89.00      Muy alto           6.64           4.60  
...     ...           ...            ...            ...  
1145    NaN           NaN           4.72           1.54  
1146    NaN           NaN           1.71           4.59  
1147    NaN           NaN           4.48           6.92  
1148    NaN           NaN           6.91           5.41  
1149    NaN           NaN           6.60           6.06  

[1150 rows x 13 columns]

In [ ]:
display(df_grades_left)

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
995     Richard Young    male        group D                 high school   
996     Wanda Russell  female        group B                 high school   
997    Marina Zeigler  female        group C           bachelor's degree   
998     Laurie Carter  female        group B            some high school   
999      Amanda Perez  female        group A                 high school   

            lunch test preparation course math score reading score  \
0        standard                    none         37            46   
1        standard               completed         80            78   
2        standard                    none         73            66   
3        standard               completed         81            75   
4        standard               completed         94            86   
..            ...                     ...        ...           ...   
995      standard                    none         69            75   
996  free/reduced               completed         23            44   
997  free/reduced               completed         66            74   
998      standard               completed         54            61   
999      standard               completed         68            80   

    writing score    GPA GPA cuantiles  science score  history score  
0              46  43.00      Bajísimo           2.50            3.8  
1              81  79.67          Alto           4.36            5.2  
2              63  67.33         Medio           4.48            3.8  
3              76  77.33          Alto           5.32            5.0  
4              87  89.00      Muy alto           6.64            4.6  
..            ...    ...           ...            ...            ...  
995            71  71.67         Medio           4.72            4.2  
996            36  34.33      Bajísimo           1.96            3.6  
997            81  73.67          Alto           4.54            4.1  
998            62  59.00          Bajo           3.58            4.3  
999            76  74.67          Alto           4.66            5.3  

[1000 rows x 13 columns]

In [ ]:
display(df_grades_right) 

names  gender race/ethnicity parental level of education  \
0           Pam Otoole  female        group B           bachelor's degree   
1          Diane Olson  female        group C                some college   
2       Sarah Flannery  female        group B             master's degree   
3     Stanley Chappell    male        group A          associate's degree   
4        Arnold Dodson    male        group C                some college   
...                ...     ...            ...                         ...   
1145     Carson Brooks     NaN            NaN                         NaN   
1146         Ruby Reed     NaN            NaN                         NaN   
1147    Zachary Torres     NaN            NaN                         NaN   
1148       Eliana Gray     NaN            NaN                         NaN   
1149       Gavin James     NaN            NaN                         NaN   

             lunch test preparation course math score reading score  \
0         standard                    none         72            72   
1         standard               completed         69            90   
2         standard                    none         90            95   
3     free/reduced                    none         47            57   
4         standard                    none         76            78   
...            ...                     ...        ...           ...   
1145           NaN                     NaN        NaN           NaN   
1146           NaN                     NaN        NaN           NaN   
1147           NaN                     NaN        NaN           NaN   
1148           NaN                     NaN        NaN           NaN   
1149           NaN                     NaN        NaN           NaN   

     writing score    GPA GPA cuantiles  science score  history score  
0               74  72.67          Alto           4.42           4.00  
1               88  82.33      Muy alto           3.88           4.80  
2               93  92.67      Muy alto           5.32           6.10  
3               44  49.33      Bajísimo           3.82           3.70  
4               75  76.33          Alto           4.60           5.20  
...            ...    ...           ...            ...            ...  
1145           NaN    NaN           NaN           4.72           1.54  
1146           NaN    NaN           NaN           1.71           4.59  
1147           NaN    NaN           NaN           4.48           6.92  
1148           NaN    NaN           NaN           6.91           5.41  
1149           NaN    NaN           NaN           6.60           6.06  

[1150 rows x 13 columns]

In [ ]:
display(df_grades_inner) 

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
995     Richard Young    male        group D                 high school   
996     Wanda Russell  female        group B                 high school   
997    Marina Zeigler  female        group C           bachelor's degree   
998     Laurie Carter  female        group B            some high school   
999      Amanda Perez  female        group A                 high school   

            lunch test preparation course math score reading score  \
0        standard                    none         37            46   
1        standard               completed         80            78   
2        standard                    none         73            66   
3        standard               completed         81            75   
4        standard               completed         94            86   
..            ...                     ...        ...           ...   
995      standard                    none         69            75   
996  free/reduced               completed         23            44   
997  free/reduced               completed         66            74   
998      standard               completed         54            61   
999      standard               completed         68            80   

    writing score    GPA GPA cuantiles  science score  history score  
0              46  43.00      Bajísimo           2.50            3.8  
1              81  79.67          Alto           4.36            5.2  
2              63  67.33         Medio           4.48            3.8  
3              76  77.33          Alto           5.32            5.0  
4              87  89.00      Muy alto           6.64            4.6  
..            ...    ...           ...            ...            ...  
995            71  71.67         Medio           4.72            4.2  
996            36  34.33      Bajísimo           1.96            3.6  
997            81  73.67          Alto           4.54            4.1  
998            62  59.00          Bajo           3.58            4.3  
999            76  74.67          Alto           4.66            5.3  

[1000 rows x 13 columns]

De lo realizado anteriormente, se observa que tanto el merge outer como el right presentan valores NaN, lo cual no es lo ideal. Por otro lado, si bien el merge inner funciona correctamente, podría resultar en una pérdida de información debido a que solo considera la intersección entre ambas tablas. Por lo tanto, se puede concluir que el merge más adecuado a utilizar es el left, pues es el que mejor preserva la información.

# Conclusión


Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>